From: Michael Catanzaro <mcatanzaro@redhat.com>
Date: Mon, 23 Jun 2025 13:25:47 -0500
Subject: update-monitor: improve logic for deciding when to show notification

We are inappropriately considering only how long it has been since the
last update notification was displayed. If it has been more than a few
days since the last notification, we say updates are outdated. But
checking the last notification timestamp really only makes sense for
deciding whether we are nagging the user with too many notifications; we
shouldn't be checking it to decide whether the system is outdated or
not. The criteria for deciding that is really based on the time of the
last check for updates or the time that updates were last installed,
neither of which are examined currently.

This commit fixes the logic such that I will no longer be notified that
"Updates Are Out of Date" unless it has actually been 7 days since the
last check for update. I have automatic updates enabled, so I expect
GNOME Software to check for updates daily.

Additionally, we are supposed to nag the user daily if it has been more
than 7 days since the last check for updates. The current code nags if
it has been exactly 4 days or more than 7 days since the last nag, but
this doesn't seem to be based on the design. Fix that.

I also consistently consider can_download as part of should_download.

Origin: upstream, 48.4, commit:5ecc19457579789f7535e27e68c41cd9c1b9d90c
Bug: https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2856
---
 src/gs-update-monitor.c | 36 +++++++++++++++++++++++-------------
 1 file changed, 23 insertions(+), 13 deletions(-)

diff --git a/src/gs-update-monitor.c b/src/gs-update-monitor.c
index 2ef1b89..e34fb65 100644
--- a/src/gs-update-monitor.c
+++ b/src/gs-update-monitor.c
@@ -200,27 +200,38 @@ should_notify_about_pending_updates (GsUpdateMonitor *monitor,
 {
 	gboolean has_important = FALSE, all_downloaded = FALSE, any_downloaded = FALSE;
 	gboolean should_download, res = FALSE;
-	gint64 timestamp_days;
+	gint64 notification_timestamp_days;
 
-	if (!get_timestamp_difference_days (monitor, "update-notification-timestamp", &timestamp_days)) {
+	if (!get_timestamp_difference_days (monitor, "update-notification-timestamp", &notification_timestamp_days)) {
 		/* Large-enough number to succeed for the initial test */
-		timestamp_days = 365;
+		notification_timestamp_days = 365;
 	}
 
-	should_download = should_download_updates (monitor);
+	/* can_download reflects whether GNOME Software should *automatically*
+	 * download updates, not whether the user is able to do so. E.g. it is
+	 * false if power saving mode is enabled.
+	 *
+	 * Beware it is expected to be spuriously false, e.g. immediately after
+	 * resuming from suspend when no network is available for a second or
+	 * two. This is OK because it should not be spuriously false for
+	 * long periods of time.
+	 */
+	should_download = should_download_updates (monitor) && can_download;
+
 	if (apps != NULL)
 		check_updates_kind (apps, &has_important, &all_downloaded, &any_downloaded);
 
 	if (apps == NULL || !gs_app_list_length (apps)) {
-		/* Notify only when the download is disabled, or cannot download, and it's the 4th day or it's more than 7 days */
-		if ((!should_download || !can_download) && (timestamp_days >= 7 || timestamp_days == 4) &&
-		     gs_plugin_loader_get_allow_updates (monitor->plugin_loader)) {
+		if (!should_download &&
+		    gs_plugin_loader_get_allow_updates (monitor->plugin_loader) &&
+		    notification_timestamp_days >= 1 &&
+		    check_if_timestamp_more_than_days_ago (monitor, "check-timestamp", 7)) {
 			*out_title = _("Updates Are Out of Date");
 			*out_body = _("Please check for available updates");
 			res = TRUE;
 		}
 	} else if (has_important) {
-		if (timestamp_days >= 1) {
+		if (notification_timestamp_days >= 1) {
 			if (all_downloaded) {
 				*out_title = _("Critical Updates Ready to Install");
 				*out_body = _("Install critical updates as soon as possible");
@@ -232,14 +243,13 @@ should_notify_about_pending_updates (GsUpdateMonitor *monitor,
 			}
 		}
 	} else if (all_downloaded) {
-		if (timestamp_days >= 3) {
+		if (notification_timestamp_days >= 3) {
 			*out_title = _("Updates Ready to Install");
 			*out_body = _("Software updates are ready and waiting");
 			res = TRUE;
 		}
-	/* To not hide downloaded updates for 14 days when new updates were discovered meanwhile.
-	   Never show "Available to Download" when it's supposed to download the updates. */
-	} else if (!should_download && timestamp_days >= 14) {
+	} else if (!should_download && notification_timestamp_days >= 3 &&
+		   check_if_timestamp_more_than_days_ago (monitor, "install-timestamp", 14)) {
 		*out_title = _("Updates Available to Download");
 		*out_body = _("Software updates can be downloaded");
 		res = TRUE;
@@ -247,7 +257,7 @@ should_notify_about_pending_updates (GsUpdateMonitor *monitor,
 
 	g_debug ("%s: last_test_days:%" G_GINT64_FORMAT " n-apps:%u should_download:%d can_download:%d has_important:%d "
 		"all_downloaded:%d any_downloaded:%d res:%d%s%s%s%s", G_STRFUNC,
-		timestamp_days, apps == NULL ? 0 : gs_app_list_length (apps), should_download, can_download, has_important,
+		notification_timestamp_days, apps == NULL ? 0 : gs_app_list_length (apps), should_download, can_download, has_important,
 		all_downloaded, any_downloaded, res,
 		res ? " reason:" : "",
 		res ? *out_title : "",
